addons/account/models/account_reconcile_model.py

```diff
@@ -636,64 +623,87 @@ def _get_invoice_matching_query(self, st_lines_with_partner, excluded_ids):
             query += 'AND aml.id NOT IN %(excluded_aml_ids)s'
             params['excluded_aml_ids'] = tuple(excluded_ids)
 
-        # Oldest due dates come first.
-        query += ' ORDER BY aml_date_maturity, aml_id'
+        if self.matching_order == 'new_first':
+            query += ' ORDER BY aml_date_maturity DESC, aml_id DESC'
+        else:
+            query += ' ORDER BY aml_date_maturity ASC, aml_id ASC'
 
         return query, params
 
     def _get_select_communication_flag(self):
-        return r'''
-            -- Determine a matching or not with the statement line communication using the aml.name, move.name or move.ref.
-            (
-                aml.name IS NOT NULL
-                AND
-                substring(REGEXP_REPLACE(aml.name, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*') != ''
-                AND
-                    regexp_split_to_array(substring(REGEXP_REPLACE(aml.name, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*'),'\s+')
-                    && regexp_split_to_array(substring(REGEXP_REPLACE(st_line.payment_ref, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*'), '\s+')
-            )
-            OR
-                regexp_split_to_array(substring(REGEXP_REPLACE(move.name, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*'),'\s+')
-                && regexp_split_to_array(substring(REGEXP_REPLACE(st_line.payment_ref, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*'), '\s+')
-            OR
-            (
-                move.ref IS NOT NULL
-                AND
-                substring(REGEXP_REPLACE(move.ref, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*') != ''
-                AND
-                    regexp_split_to_array(substring(REGEXP_REPLACE(move.ref, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*'),'\s+')
-                    && regexp_split_to_array(substring(REGEXP_REPLACE(st_line.payment_ref, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*'), '\s+')
-            )                                   AS communication_flag
-        '''
+        self.ensure_one()
+        # Determine a matching or not with the statement line communication using the aml.name, move.name or move.ref.
+        st_ref_list = []
+        if self.match_text_location_label:
+            st_ref_list += ['st_line.payment_ref']
+        if self.match_text_location_note:
+            st_ref_list += ['st_line_move.narration']
+        if self.match_text_location_reference:
+            st_ref_list += ['st_line_move.ref']
+
+        st_ref = " || ' ' || ".join(
+            "COALESCE(%s, '')" % st_ref_name
+            for st_ref_name in st_ref_list
+        )
+        if not st_ref:
+            return "FALSE"
+
+        statement_compare = r"""(
+                {move_field} IS NOT NULL AND substring(REGEXP_REPLACE({move_field}, '[^0-9\s]', '', 'g'), '\S(?:.*\S)*') != ''
```

Comment:
just for information, in PostgreSQL, you can use double quotes around table names and field names. I think it's safer.